extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/test/CalciteSelectQueryTestMSQ.java

```diff
@@ -0,0 +1,369 @@
+
+      @Override
+      public Interval getDataInterval()
+      {
+        return segmentId.getInterval();
+      }
+
+      @Nullable
+      @Override
+      public QueryableIndex asQueryableIndex()
+      {
+        return index;
+      }
+
+      @Override
+      public StorageAdapter asStorageAdapter()
+      {
+        return new QueryableIndexStorageAdapter(index);
+      }
+
+      @Override
+      public void close()
+      {
+      }
+    };
+    return new Supplier<Pair<Segment, Closeable>>()
+    {
+      @Override
+      public Pair<Segment, Closeable> get()
+      {
+        return new Pair<>(segment, Closer.create());
+      }
+    };
+  }
+
+  protected Map<String, Object> defaultScanQueryContext(final RowSignature signature)
+  {
+    try {
+      return ImmutableMap.<String, Object>builder()
+                         .putAll(MSQTestBase.DEFAULT_MSQ_CONTEXT)
+                         .put(
+                             DruidQuery.CTX_SCAN_SIGNATURE,
+                             queryFramework().queryJsonMapper().writeValueAsString(signature)
+                         )
+                         .build();
+    }
+    catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+
+
+  @Test
+  public void testOrderThenLimitThenFilter()
+  {
+    RowSignature resultSignature = RowSignature.builder()
+                                               .add("dim1", ColumnType.STRING)
+                                               .build();
+
+    testQueryWithMSQ(
+        "SELECT dim1 FROM "
+        + "(SELECT __time, dim1 FROM druid.foo ORDER BY __time DESC LIMIT 4) "
+        + "WHERE dim1 IN ('abc', 'def')",
+        ImmutableList.of(
+            newScanQueryBuilder()
+                .dataSource(
+                    new QueryDataSource(
+                        newScanQueryBuilder()
+                            .dataSource(CalciteTests.DATASOURCE1)
+                            .intervals(querySegmentSpec(Filtration.eternity()))
+                            .columns(ImmutableList.of("__time", "dim1"))
+                            .limit(4)
+                            .order(ScanQuery.Order.DESCENDING)
+                            .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
+                            .context(defaultScanQueryContext(resultSignature))
+                            .build()
+                    )
+                )
+                .intervals(querySegmentSpec(Filtration.eternity()))
+                .columns(ImmutableList.of("dim1"))
+                .filters(in("dim1", Arrays.asList("abc", "def"), null))
+                .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
+                .context(defaultScanQueryContext(resultSignature))
+                .build()
+        ),
+        ImmutableList.of(
+            new Object[]{"abc"},
+            new Object[]{"def"}
+        ),
+        new MSQTestBase.ExtractResults(indexingServiceClient)
```

Comment:
Getting the results seems pretty generic: the kind of thing that should be built into the query framework, not specified per-test.

Since the test framework will need the indexing service client, one way to provide that is via the test config class which is passed into the builder, then onto the runner.

Reply:
Updated